File: antirobot/lib/hypocrisy/chacha.cpp
#include "chacha.h"

#include <util/stream/output.h>


namespace NHypocrisy {


namespace {


ui32 Rotl(ui32 operand, ui32 shift) {
    return (operand << shift) | (operand >> (32 - shift));
}


void PerformQuarterRound(ui32* dst, size_t a, size_t b, size_t c, size_t d) {
    dst[a] += dst[b];
    dst[d] = Rotl(dst[d] ^ dst[a], 16);

    dst[c] += dst[d];
    dst[b] = Rotl(dst[b] ^ dst[c], 12);

    dst[a] += dst[b];
    dst[d] = Rotl(dst[d] ^ dst[a], 8);

    dst[c] += dst[d];
    dst[b] = Rotl(dst[b] ^ dst[c], 7);
}


} // namespace


void PerformChaChaDoubleRound(ui32* dst) {
    PerformQuarterRound(dst, 0, 4, 8, 12);
    PerformQuarterRound(dst, 1, 5, 9, 13);
    PerformQuarterRound(dst, 2, 6, 10, 14);
    PerformQuarterRound(dst, 3, 7, 11, 15);

    PerformQuarterRound(dst, 0, 5, 10, 15);
    PerformQuarterRound(dst, 1, 6, 11, 12);
    PerformQuarterRound(dst, 2, 7, 8, 13);
    PerformQuarterRound(dst, 3, 4, 9, 14);
}


} // namespace NHypocrisy

File: antirobot/lib/hypocrisy/tools/prepare/chacha.js
const CHACHA_QUARTER_ROUNDS = [
    [0, 4, 8, 12],
    [1, 5, 9, 13],
    [2, 6, 10, 14],
    [3, 7, 11, 15],
    [0, 5, 10, 15],
    [1, 6, 11, 12],
    [2, 7, 8, 13],
    [3, 4, 9, 14]
];

function mix(params, numRounds) {
    let output = Array.from(params);

    for (let i = 0; i < numRounds; i += 1) {
        performDoubleRound(output);
    }

    for (let i = 0; i < output.length; i += 1) {
        output[i] = (output[i] + params[i]) >>> 0;
    }

    return output;
}

function performDoubleRound(output) {
    for (const quarterRound of CHACHA_QUARTER_ROUNDS) {
        performQuarterRound(output, ...quarterRound);
    }
}

function performQuarterRound(output, a, b, c, d) {
    output[a] += output[b];
    output[d] = rotl(output[d] ^ output[a], 16);

    output[c] += output[d];
    output[b] = rotl(output[b] ^ output[c], 12);

    output[a] += output[b];
    output[d] = rotl(output[d] ^ output[a], 8);

    output[c] += output[d];
    output[b] = rotl(output[b] ^ output[c], 7);

    output[a] >>>= 0;
    output[b] >>>= 0;
    output[c] >>>= 0;
    output[d] >>>= 0;
}

function rotl(operand, shift) {
    return ((operand << shift) | (operand >>> (32 - shift))) >>> 0;
}

module.exports = {
    CHACHA_QUARTER_ROUNDS,
    mix,
    performDoubleRound,
    performQuarterRound
}
